pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java

```diff
@@ -556,13 +618,90 @@ public synchronized void redeliverUnacknowledgedMessages(Consumer consumer) {
+                        @Override
+                        public void readEntryComplete(Entry entry, Object ctx) {
+                            if (entry == null) {
+                                log.error("[{}-{}] Read an null entry from cursor {}", name, consumer, position);
+                                latch.countDown();
+                            } else {
+                                try {
+                                    ByteBuf headersAndPayload = entry.getDataBuffer();
+                                    MessageImpl<byte[]> msg = MessageImpl.deserialize(headersAndPayload);
+                                    headersAndPayload.retain();
+                                    msg.setReplicatedFrom("DLQ");
+                                    CompletableFuture<MessageId> future = deadLetterTopicProducer.sendAsync(msg);
+                                    future.whenCompleteAsync((messageId, error) -> {
+                                        if (error != null) {
+                                            log.error("[{}-{}] Fail to send message to dead letter topic {} {} {}",
+                                                    name, consumer, deadLetterTopic, error.getMessage(), error);
+                                            messagesToReplay.add(position.getLedgerId(), position.getEntryId());
+                                            latch.countDown();
+                                        } else {
+                                            cursor.asyncDelete(position, deleteCallback, position);
+                                            redeliveryTracker.remove(position);
+                                            latch.countDown();
+                                        }
+                                    });
+                                } catch (Throwable t) {
+                                    log.error("[{}-{}] Failed to deserialize message at {} {} {} {}", name, consumer,
+                                            entry.getPosition(), entry.getLedgerId(), t.getMessage(), t);
+                                    cursor.asyncDelete(position, deleteCallback, position);
+                                    redeliveryTracker.remove(position);
+                                    entry.release();
+                                    latch.countDown();
+                                }
+                            }
+                        }
+                        @Override
+                        public void readEntryFailed(ManagedLedgerException exception, Object ctx) {
+                            log.error("[{}-{}] Read entries failed {} {}", name, consumer, exception.getMessage(), exception);
+                            messagesToReplay.add(position.getLedgerId(), position.getEntryId());
+                            latch.countDown();
+                        }
+                    }, null);
+                }
+                try {
```

Comment:
I don't think we need to wait for the results here. because the logic has been processed in the callbacks. right?

Reply:
Has a problem with messagesToReplay
```java
Set<PositionImpl> messagesToReplayNow = messagesToReplay.items(messagesToRead).stream()
                        .map(pair -> new PositionImpl(pair.first, pair.second)).collect(toSet());

                if (log.isDebugEnabled()) {
                    log.debug("[{}] Schedule replay of {} messages for {} consumers", name, messagesToReplayNow.size(),
                            consumerList.size());
                }

                havePendingReplayRead = true;
                Set<? extends Position> deletedMessages = cursor.asyncReplayEntries(messagesToReplayNow, this,
                        ReadType.Replay);
                // clear already acked positions from replay bucket

                deletedMessages.forEach(position -> messagesToReplay.remove(((PositionImpl) position).getLedgerId(),
                        ((PositionImpl) position).getEntryId()));
```

Reply:
No sure I understand the problem here. Can you clarify it?

Reply:
Oh, I thought that if the last redeliverUnacknowledgedMessages is return but callbacks not complete yet, then another redeliverUnacknowledgedMessages is coming,  if last messagesToReplay is not delete yet, so that current messagesToReplay has the last messagesToReplay items.
And, I thought that call readMoreEntries() should wait for callbacks complete.